### PR TITLE
Assignment4

### PR DESCRIPTION
**Pull request by CarlyLanglois**
SQL tasks are added now

### CODE REVIEW

File: src/assignments/tech-jobs-persistent.rst [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+you'll need to do the work to connect the project to a database for storing user-submitted job data.
+
+Checkout and Review the Starter Code
+------------------------------------
+
+Fork and clone the starter code from the `techjobs persistent repository <https://github.com/LaunchCodeEducation/java-web-dev-techjobs-persistent>`__.
+
+You won't be able to run your application yet. You'll need to complete :ref:`Part1 <tech-jobs-persistent-pt1>` before you can
+check out the app in the browser. That said, it's a good idea to get familiar with the classes and templates even before you're able
+to execute ``bootRun``.
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: I think it would be a good idea to include an example of what kind of errors they might get when they open their starter code. That way they have some sort of orientation about what to expect from the starter code. If that might be awkward, perhaps encourage them to check out a particular class, first?

File: src/assignments/tech-jobs-persistent.rst [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+   
+      Remember to double click on the schema name in the file tree to make it the default schema.
+
+#. In the administration tab, create a new user, ``techjobs`` with the same settings as described in
+   the lesson tutorial.
+   TODO: reference the point in the book setting up the new user.
+
+#. Update ``build.gradle`` with the dependencies.
+
+#. Update ``src/resources/application.properties`` with the right info.
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: I would add some additional info to these steps. Maybe how many dependencies or give the general names of the dependencies, like MySQL. Or the "... with the right database info."

File: src/assignments/tech-jobs-persistent.rst [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+.. admonition:: Tip
+   
+   You can double check your setup against what you've already done for 
+   :ref:`your coding events repo <setup-orm-database>`
+
+Test It!
+^^^^^^^^
+
+When your database is properly configured, you should have no compiler errors when starting the application. Execute ``bootRun``
+and check the compiler output to make sure this is the case. If all runs, you will be able to view your app running locally in
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: "If all runs, ..." is a little awkward. Perhaps, "If everything runs, ..."?

File: src/assignments/tech-jobs-persistent.rst [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+   :ref:`your coding events repo <setup-orm-database>`
+
+Test It!
+^^^^^^^^
+
+When your database is properly configured, you should have no compiler errors when starting the application. Execute ``bootRun``
+and check the compiler output to make sure this is the case. If all runs, you will be able to view your app running locally in 
+the browser at ``Localhost:8080`` (unless of course you have changed the server port).
+
+Your running application still has limited functionality. You won't yet be able to add a job from the *Add Job* form. You also
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: "add a job with the *Add Job* form."

File: src/assignments/tech-jobs-persistent.rst [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+won't yet be able to view the list of jobs or search for jobs - but this is mostly because you have no more job data. Move on to
+Part 2 below to get the app to be more useful.
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: " below to start adding these functionalities."

File: src/assignments/tech-jobs-persistent.rst [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+---------------------------------------
+
+You will need to have completed the :ref:`setup steps <tech-jobs-persistent-setup>` before starting this
+section.
+
+``AbstractEntity``
+^^^^^^^^^^^^^^^^^^
+
+We've replaced the abstract class ``JobField`` with an even more abstracted class aptly named, 
+``AbstractEntity``. This class is holds the fields and methods that are common across the ``Job`` class
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: "This class holds the fields and methods... "

File: src/assignments/tech-jobs-persistent.rst [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+
+#. Since all of the subclasses of ``AbstractEntity`` will be entities themselves, add the ``@Id`` 
+   and ``@GeneratedValue`` annotations to the field ``id``.
+
+#. Each subclass will also inherit the ``name`` field from ``AbstractEntity``. Add appropriate 
+   validation annotations so that:
+   
+   a. a user cannot leave this field blank when creating an object. 
+
+   b. there are reasonable limitations on the size of the name string.
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: What would be reasonable? I might add some specificity to this.

File: src/assignments/tech-jobs-persistent.rst [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+
+Models
+^^^^^^
+
+In the last assignment, a ``Job`` object contained string fields for employer and core competency data. This employer 
+and skill (formerly core competency) information about a particular job will now be stored in classes themselves.
+These items themselves will hold their own supplementary information. 
+
+#. Open the ``Employer`` model class. In addition to the fields inherited from ``AbstractEntity``, ``Employer`` should have a 
+   string field for ``location``. Add this field with validation as you see fit, as well as getters and setters.
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: I think some students would decide that they don't want to add validation with this phrasing. Perhaps, "Add the field for ``location`` with validation. In addition, add getters and setters to ``Employer``."

File: src/assignments/tech-jobs-persistent.rst [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+#. Repeat the steps above for an interface, ``SkillRepository``.
+
+Controllers
+^^^^^^^^^^^
+
+With the employer repository in place, we will reference this to send object information through 
+the ``EmployerController`` handlers. ``EmployerController`` contains two handlers with missing 
+information.
+
+#. View an individual employer object.
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: This is a little disjointed.

File: src/assignments/tech-jobs-persistent.rst [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+
+#. Create a ``SkillController`` class and replicate the steps you followed above for ``EmployerController``.
+
+Test It!
+^^^^^^^^
+The employer and skill view templates for adding and viewing these objects are made for you. Before you move on,
+test your application now to make sure it runs as expected. You should be able to create Employer and Skill objects
+and view them.
+
+#. Start up your application – don’t forget to have our SQL server running – and go to the *Add Jobs*
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: "... have your SQL server... "

File: src/assignments/tech-jobs-persistent.rst [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+   If everything seems to work – that is, you are
+   able to submit the form without any errors – but you don’t see your
+   employers or skills in the list after submission, here’s what you should check:
+
+   #. Is there any data in the ``employers`` and ``skills`` table? Check by going to MySQL Workbench
+      and looking for the employer/skill data within your schema.
+
+   #. If there’s data in the database, check that you are correctly
+      querying for the list of all objects in the controller
+      findAll()``?
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: Syntax highlighting is off here.

File: src/assignments/tech-jobs-persistent.rst [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+   will represent the list of all items in a given job. We’ll do this
+   in a bit.
+
+#. Use the ``@OneToMany`` and ``@JoinColumn`` annotations on the jobs list in ``Employer`` to declare the relationship between   
+   data tables.
+
+Update ``Job`` Model
+^^^^^^^^^^^^^^^^^^^^
+
+#. Since it too has ``id`` and ``name`` fields, the ``Job`` model class can also inherit from ``AbstractEntity``. Update the
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: "Since the ``Job`` model class has ``id`` and ``name``...., it too can inherit from ``AbstractEntity``."

File: src/assignments/tech-jobs-persistent.rst [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+#. Add a field ``employerRepository`` annotated with ``@Autowired``.
+#. A user will select an employer when they create a job. Add the employer data from ``employerRepository`` into the form template.
+   The add job form already includes an employer selection option. Be sure your variable name for the employer data matches that 
+   already used in ``templates/add``. 
+#. Checkout ``templates/add.html``. Make a mental note of the name of the variable being used to pass the selected employer 
+   id on form submission.
+#. In ``processAddJobForm``, add a parameter to the method to pass in the template variable you just found. You'll need to use the 
+   ``@RequestParam`` annotation on this parameter. 
+#. Still in ``processAddJobForm``, add code inside of this method to select the employer object that has been chosen to be 
+   affiliated with the new job. You will need to select the employer useing the request parameter you've added to the method.
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: "using"

File: src/assignments/tech-jobs-persistent.rst [hunk 14]
```diff
@@ -0,0 +1,382 @@
...
+
+   #. This field has a many-to-many type relationship with skills. You'll need to add the ``@ManyToMany`` annotation 
+      with an argument ``mappedBy="skills"`` to ensure this mapping.
+
+Refactor ``Job.skills``
+^^^^^^^^^^^^^^^^^^^^^^^
+
+#. Update your ``Job`` model class to fit its many-to-many relationship with skills.
+
+   #. ``Job.skills`` already exists. What needs to change and/or be added to map this relationship.
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: End this point with a question mark.

**Review by gildedgardenia — COMMENTED** [review]
Overall, a great assignment! I made some comments on minor typos and phrasing things.

File: src/assignments/tech-jobs-persistent.rst (hunk 9 above)
Comment (CarlyLanglois) [reply src/assignments/tech-jobs-persistent.rst]: yeah i dont really know why thats there! going to consider it a typo

File: src/assignments/tech-jobs-persistent.rst [hunk 15]
```diff
@@ -0,0 +1,414 @@
...
+Spring Data annotations and classes. Some of these have already been added but the dependency has not yet been defined. 
+That will be one of your tasks. You'll need to complete :ref:`Part1 <tech-jobs-persistent-pt1>` before you can
+check out the project running in the browser. 
+
+That said, it's a good idea to scan the classes and templates even before you're able to execute 
+``bootRun``. Take a gander at the ``Job`` class. It will look somewhat similar to the model in 
+:ref:`Tech Jobs (MVC Edition) <tech-jobs-mvc>`, with a few key differences. 
+
+We're no longer using a csv file to load job data, instead, we'll be creating new Job objects via a 
+user form. The Job data will be stored in a MySQL database that you'll setup in :ref:`Part 1 <tech-jobs-persistent-pt1>` of this assignment.
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: Switch "we" to "you"

File: src/assignments/tech-jobs-persistent.rst [hunk 16]
```diff
@@ -0,0 +1,414 @@
...
+
+You'll next be tasked with completing the work to persist some of the classes of the job data. As you explore
+the starter code, you'll notice that the ``JobField`` abstract class is no longer present. Your task for 
+:ref:`Part 2 <tech-jobs-persistent-pt2>` is to complete the work to persist some of the classes.
+You'll do this for ``Employer`` and ``Skill`` classes, as well as ``Job``.
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: The topic and third sentence of this paragraph come off as redundant. Maybe rephrase?

File: src/assignments/tech-jobs-persistent.rst [hunk 17]
```diff
@@ -0,0 +1,414 @@
...
+
+The ``Job`` class will also look different from how you have last seen it. In Parts 3 and 4, you'll 
+add object relational mapping on the ``Job`` class by refactoring the ``employer`` and ``skills`` (formerly ``coreCompetency``) 
+fields. 
+
+In your IntelliJ project, you'll see an empty file in the root directory called ``queries.sql``. After completing the 
+Java updates for parts 1,2,3, and 4, we ask you to test your application updates with SQL statements. Since you are entering
+your own data, running the queries we ask you to write may not return any results. However, as the architect of the database,
+you will have the knowledge to write these nonetheless.
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: This is a little awkward. Maybe, "Since you are entering your own data, the queries we ask you to write will return unique result sets. For example, if you haven't entered any data yet, there may be an empty result set! However, as the architect of the database, you have the knowledge to write the appropriate queries nonetheless."

File: src/assignments/tech-jobs-persistent.rst [hunk 18]
```diff
@@ -0,0 +1,414 @@
...
+
+#. Start MySQL Workbench and create a new schema named ``techjobs``.
+
+   .. admonition:: Tip
+   
+      Remember to double click on the schema name in the file tree to make it the default schema.
+
+#. In the administration tab, create a new user, ``techjobs`` with the same settings as described in
+   the lesson tutorial.
+   TODO: reference the point in the book setting up the new user.
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: This is covered at the same point as the rest of the database setup steps.

File: src/assignments/tech-jobs-persistent.rst [hunk 19]
```diff
@@ -0,0 +1,414 @@
...
+Java updates for parts 1,2,3, and 4, we ask you to test your application updates with SQL statements. Since you are entering
+your own data, running the queries we ask you to write may not return any results. However, as the architect of the database,
+you will have the knowledge to write these nonetheless.
+
+.. _tech-jobs-persistent-pt1:
+
+Part 1: Setup the Database
+--------------------------
+
+Connect a database to a Spring App.
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: I would take this out and make it the title of the part in the heading.

File: src/assignments/tech-jobs-persistent.rst [hunk 20]
```diff
@@ -0,0 +1,414 @@
...
+
+   .. admonition:: Tip
+   
+      Remember to double click on the schema name in the file tree to make it the default schema.
+
+#. In the administration tab, create a new user, ``techjobs`` with the same settings as described in
+   the lesson tutorial.
+   TODO: reference the point in the book setting up the new user.
+
+#. Update ``build.gradle`` with the dependencies.
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: "the necessary dependencies"

File: src/assignments/tech-jobs-persistent.rst [hunk 21]
```diff
@@ -0,0 +1,414 @@
...
+and check the compiler output to make sure this is the case. If everything runs, you will be able to view your app 
+locally in the browser at ``Localhost:8080`` (unless of course you have changed the server port).
+
+#. In your MySQL workbench, open a new query tab to check your database connection.
+
+#. **SQL TASK:** At this point, you will have one table, ``job``. In ``queries.sql`` under "Part 1", list the columns and their data types 
+   in the table.
+
+Your running application still has limited functionality. You won't yet be able to add a job with the *Add Job* form. You also
+won't yet be able to view the list of jobs or search for jobs - but this is mostly because you have no more job data. Move on to
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: "you have no job data"

File: src/assignments/tech-jobs-persistent.rst [hunk 22]
```diff
@@ -0,0 +1,414 @@
...
+#. As with ``Employer``, give this class the ``@Entity`` annotation and be sure it contains a no-arg
+   constructor.
+
+
+Data Layer
+^^^^^^^^^^
+
+To map the ``Employer`` and ``Skill`` classes to your techjobs database, you'll add data access interfaces for these relational 
+objects, similiar to the existing ``JobRepository`` interface. Like ``JobRepository``, make use of the Spring Data 
+``CrudRepository`` class to map our objects.
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: ``CrudRepository`` is an interface.

File: src/assignments/tech-jobs-persistent.rst [hunk 23]
```diff
@@ -0,0 +1,414 @@
...
+information. Your task here is to make use of the ``EmployerRepository`` class in these handlers. 
+
+#. Add a private field of ``EmployerRepository`` type called ``employerRepository`` to 
+   ``EmployerController``. Give this field an ``@Autowired`` annotation.
+
+#. ``processAddEmployerForm`` already takes care of sending the form back if any of the submitted 
+   employer object information is invalid. However, it doesn't yet contain the code to save a
+   valid object. Use ``employerRepository`` and the appropriate method to do so.
+   
+#. ``displayViewEmployer`` will be in charge of rendering an a page to view the contents of an individual
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: "an a"

File: src/assignments/tech-jobs-persistent.rst [hunk 24]
```diff
@@ -0,0 +1,414 @@
...
+clean by removing the job table. It will be recreated when you run the application again.
+   
+#. **SQL TASK:** In ``queries.sql`` under "Part 3", write the SQL statement to remove the job table.
+
+
+The *List* and *Search* functionality still isn't quite fixed so to view a job in the application, make a note 
+of the job's id in the SQL table. Back in your browser, enter the path for ``/view/{jobId}``.
+
+
+When everything works, move on to Part 3 below.
```
Comment (gildedgardenia) [inline src/assignments/tech-jobs-persistent.rst]: "Part 4"

**Review by gildedgardenia — APPROVED** [review]
Some minor changes, but I love the assignment!
